### DISCUSSION TITLE
Using fpp types in other components

### DISCUSSION THREAD

**Discussion by pauladam316** [nasa/fprime#1921] [Q&A]
Hello!

I'm having dependency issues with a slightly off-normal setup and looking for guidance. I have a struct type that I define in an fpp file for component A, that has no accompanying cpp or hpp file. I then have component B that declares an instance of the struct defined in component A in its cpp file. If I try to use it I get errors that it cannot be found. I've tried two ways of making this work

1. in component B, I include the header for component A that is generated as part of the build process. This doesn't work well as the file has to already exist, and if you are making a fresh build it will error out trying to include a file that hasn't yet been generated
2. in component B, making a dummy port that uses the struct from port A. This seems to work, but is a weird hack

Is there a proper way to do this with F-prime? Or will I have to find my own way to make it work. I thought the MOD_DEPS flag in CmakeLists.txt would be for this very purpose, but it doesn't seem to solve the issue

**Comment by LeStarch** [comment]
Which errors are occurring? Cannot find type (include error), missing symbol (linker error), or an FPP undefined symbol (FPP error)?

You probably need to include the serializable header: `#include <path/to/component-a/[StructTypeName]SerializableAc.hpp>`.  The struct will be autocoded into a separate header from the component A where it is defined.

**Comment by bocchino** [comment]
I think the solution is (1) to add the header include as Michael said and (2) use MOD_DEPS to express the build dependency of build module B on build module A (i.e., in the CMakeLists.txt for build module B, add `set(MOD_DEPS A)`). If you do (1) but not (2) then the build may work, or you may get a linker error, or you may get a compiler error reporting an undefined header file as you described.

> in component B, making a dummy port that uses the struct from port A.

I think this works because when there is an FPP dependency from build module B to build module A, all the header inclusions, autocoder dependencies, and linker dependencies are automatically generated. When there's no FPP dependency, you have to manually add the header include and the build module dependency. The CMake build system is not currently able to build module dependencies from C++ header dependencies.
